tools/f2fsdist.py

```diff
@@ -0,0 +1,231 @@
+    u64 ts=bpf_ktime_get_ns();
+    start.update(&tid,&ts);
+    return 0;
+}
+F2FS_TRACE_READ_CODE
+static int trace_return(struct pt_regs *ctx, const char *op)
+{
+    u64 *tsp;
+    u64 pid_tgid=bpf_get_current_pid_tgid();
+    u32 pid=pid_tgid>>32;
+    u32 tid=(u32)pid_tgid;
+    //fetch timestamp and calculate delta
+    tsp=start.lookup(&tid);
+    if(tsp==0)
+    {
+        return 0; //missed start or filtered
+    }
+    u64 delta=bpf_ktime_get_ns() - *tsp;
+    start.delete(&tid);
+
+    //Skip entries with backwards time: temp workarount for
+    if((s64) delta <0)
+        return 0;
+    delta/=FACTOR;
+
+    //store as histogram
+    dist_key_t key={.slot = bpf_log2l(delta)};
+    __builtin_memcpy(&key.op, op, sizeof(key.op));
+    dist.increment(key);
+
+    return 0;
+}
+
+int trace_read_return(struct pt_regs *ctx)
+{
+    char *op = "read";
+    return trace_return(ctx,op);
+}
+int trace_write_return(struct pt_regs *ctx)
+{
+    char *op = "write";
+    return trace_return(ctx,op);
+}
+int trace_open_return(struct pt_regs *ctx)
+{
+    char *op = "open";
+    return trace_return(ctx,op);
+}
+int trace_fsync_return(struct pt_regs *ctx)
+{
+    char *op = "fsync";
+    return trace_return(ctx,op);
+}
+"""
+# Starting from Linux 4.10 f2fs_file_operations.read_iter has been changed from
+# using generic_file_read_iter() to its own f2fs_file_read_iter().
+#
+# To detect the proper function to trace check if f2fs_file_read_iter() is
+# defined in /proc/kallsyms, if it's defined attach to that function, otherwise
+# use generic_file_read_iter() and inside the trace hook filter on f2fs read
+# events (checking if file->f_op == f2fs_file_operations).
+
+if BPF.get_kprobe_functions(b'f2fs_file_read_iter'):
+    f2fs_read_fn = 'f2fs_file_read_iter'
+    f2fs_trace_read_fn = 'trace_entry'
+    f2fs_trace_read_code = ''
+else:
+    f2fs_read_fn = 'generic_file_read_iter'
+    f2fs_trace_read_fn = 'trace_read_entry'
+    f2fs_file_ops_addr = ''
+    with open(kallsyms) as syms:
+        for line in syms:
+            (addr, size, name) = line.rstrip().split(" ", 2)
+            name = name.split("\t")[0]
+            if name == "f2fs_file_operations":
+                f2fs_file_ops_addr = "0x" + addr
+                break
+        if f2fs_file_ops_addr == '':
+            print("ERROR: no f2fs_file_operations in /proc/kallsyms. Exiting.")
+            print("HINT: the kernel should be built with CONFIG_KALLSYMS_ALL.")
+            exit()
+
+    f2fs_trace_read_code = """
+int trace_read_entry(struct pt_regs *ctx, struct kiocb *iocb)
+{
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = (u32)pid_tgid;
+
+    if (FILTER_PID)
+        return 0;
+
+    // f2fs filter on file->f_op == f2fs_file_operations
+    struct file *fp = iocb->ki_filp;
+    if ((u64)fp->f_op != %s)
+        return 0;
+
+    u64 ts = bpf_ktime_get_ns();
+    start.update(&tid, &ts);
+    return 0;
+}""" % f2fs_file_ops_addr
```

Comment:
f2fs_file_ops_addr is an address in kallsyms, what if the platform's kptr_restrict is 2?